File: core/core-domain/src/main/java/com/nmh/commerce/coupon/PercentageDiscountPolicy.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.domain.Money;
import com.nmh.commerce.domain.DiscountRate;
import lombok.Getter;

@Getter
public class PercentageDiscountPolicy implements DiscountPolicy {
    private final DiscountRate discountRate;

    public PercentageDiscountPolicy(DiscountRate discountRate) {
        this.discountRate = discountRate;
    }

    @Override
    public Money calculateDiscount(Money originalPrice) {
        Money discountPrice = originalPrice.apply(discountRate);
        return originalPrice.minus(discountPrice);
    }
}

File: core/core-domain/src/main/kotlin/com/nmh/commerce/coupon/PercentageDiscountPolicy.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.domain.DiscountRate
import com.nmh.commerce.domain.Money

class PercentageDiscountPolicy(
    val discountRate: DiscountRate
) : DiscountPolicy {
    override fun calculateDiscount(originalPrice: Money): Money {
        val discountPrice = originalPrice.apply(discountRate)
        return originalPrice.minus(discountPrice)
    }
}
